### PR TITLE
changing log structure for DEV mode

### LINKED ISSUE #190: logging helpers not properly working
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current Behavior

We have a set of helpers under `controllers/utils.go` named `debug`, and `info` that allows us to log while providing different verbosities. The verbosity is in the range [0-4], the higher the number, the lower the severity. As the underlying logger, we use `zapr`. The problem of using zapr as `logr.Logger` for the controller-runtime, is that the verbosity levels have a completely different range and meaning, leading to not being able to customize the verbosity. Indeed,  using `info` from within a controller doesn't log anything.

### Expected Behavior

`info` should produce log with `info` severity, `debug` with `debug` severity, and so on. Only `debug` works at the moment because the debug level of `logr.Logger` is the same as in `zapr`.

### Steps To Reproduce

* log something through the `info` helper. No log will happen.

### Proposed solution

In the `zapr` repo README (https://github.com/go-logr/zapr#increasing-verbosity) there is a section related to adapting `zapr` verbosity levels to the `logr` ones.

### PR DESCRIPTION
**Pull request by mlavacca**
**What this PR does / why we need it**:
Log structure refactoring

**Which issue this PR fixes**
Fixes #190 

**Special notes for your reviewer**:

Logs with dev mode:
```console
2022-08-30T15:12:38.555+0200 - DEBUG - DATAPLANE - validating DataPlane resource conditions - {"namespace": "default", "name": "kong-zkr5n"}
2022-08-30T15:12:38.600+0200 - DEBUG - GATEWAY - reconciling gateway resource - {"namespace": "default", "name": "kong"}
2022-08-30T15:12:38.600+0200 - DEBUG - DATAPLANE - reconciling DataPlane resource - {"namespace": "default", "name": "kong-zkr5n"}
2022-08-30T15:12:41.441+0200 - DEBUG - GATEWAY - checking gatewayclass - {"namespace": "default", "name": "kong"}
2022-08-30T15:12:41.441+0200 - DEBUG - DATAPLANE - validating DataPlane resource conditions - {"namespace": "default", "name": "kong-zkr5n"}
2022-08-30T15:12:41.441+0200 - DEBUG - DATAPLANE - exposing DataPlane deployment via service - {"namespace": "default", "name": "kong-zkr5n"}
```

logs without dev mode:
```console
{"level":"debug","ts":1661865281.052911,"logger":"gateway","msg":"ensuring dataplane config is up to date","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"036519c4-a627-4912-91e6-476458379a5d","namespace":"default","name":"kong"}
{"level":"debug","ts":1661865281.052948,"logger":"gateway","msg":"waiting for dataplane readiness","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"036519c4-a627-4912-91e6-476458379a5d","namespace":"default","name":"kong"}
{"level":"debug","ts":1661865281.05301,"logger":"gateway","msg":"looking for associated controlplanes","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"036519c4-a627-4912-91e6-476458379a5d","namespace":"default","name":"kong"}
{"level":"debug","ts":1661865281.053045,"logger":"gateway","msg":"ensuring controlplane config is up to date","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"036519c4-a627-4912-91e6-476458379a5d","namespace":"default","name":"kong"}
{"level":"debug","ts":1661865281.053072,"logger":"gateway","msg":"waiting for controlplane readiness","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"036519c4-a627-4912-91e6-476458379a5d","namespace":"default","name":"kong"}
{"level":"debug","ts":1661865281.0530849,"logger":"gateway","msg":"ensuring DataPlane's NetworkPolicy is created","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"036519c4-a627-4912-91e6-476458379a5d","namespace":"default","name":"kong"}
```

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect significant changes

### CODE REVIEW

**Review by pmalek — COMMENTED** [review]
Do you mind adding some code snippets and/or screenshots showing the differences introduced?

File: internal/manager/run.go [hunk 1]
```diff
@@ -107,15 +109,32 @@ func DefaultConfig() Config {
...
 		ControlPlaneControllerEnabled: true,
 		DataPlaneControllerEnabled:    true,
 	}
 }
 
+// SetupEncoder sets some configurations proper to the development mode
+func SetupEncoder(developmentMode bool, options zap.Options) zap.Options {
+	if developmentMode {
+		// default to ISO8601 time encoding (e.g., 2022-08-25T14:05:51.352+0200)
+		options.TimeEncoder = zapcore.ISO8601TimeEncoder
```
[the remark this answers is not in the document]
Comment (mlavacca) [reply internal/manager/run.go:124]: What would you put into consts? `zapcore.ISO8601TimeEncoder` and stuff like that?

[the remark this answers is not in the document]
Comment (pmalek) [reply internal/manager/run.go:124]: Yup, and name it like `defaultDevTimeEncoder` or something like that. Ideally we'd have all those logger options in one place.

**Comment by pmalek** [comment]
You have conflicts to resolve :)

```diff
-{"level":"debug","ts":1661865281.0530849,"logger":"gateway","msg":"ensuring DataPlane's NetworkPolicy is created","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"036519c4-a627-4912-91e6-476458379a5d","namespace":"default","name":"kong"}
+{"level":"debug","ts":1661865281.0530849,"logger":"gateway","msg":"ensuring DataPlane's NetworkPolicy is created","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"reconcileID":"036519c4-a627-4912-91e6-476458379a5d","namespace":"default","name":"kong"}
```


It would be ideal if we could remove those duplicated `namespace` and `name` fields.

File: internal/manager/run.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mlavacca) [reply internal/manager/run.go]: I moved all these settings under `internal/manager/logging`, let me know if it works for you :)

**Comment by mlavacca** [comment]
> You have conflicts to resolve :)
> 
> ```diff
> -{"level":"debug","ts":1661865281.0530849,"logger":"gateway","msg":"ensuring DataPlane's NetworkPolicy is created","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"namespace":"default","name":"kong","reconcileID":"036519c4-a627-4912-91e6-476458379a5d","namespace":"default","name":"kong"}
> +{"level":"debug","ts":1661865281.0530849,"logger":"gateway","msg":"ensuring DataPlane's NetworkPolicy is created","controller":"gateway","controllerGroup":"gateway.networking.k8s.io","controllerKind":"Gateway","gateway":{"name":"kong","namespace":"default"},"reconcileID":"036519c4-a627-4912-91e6-476458379a5d","namespace":"default","name":"kong"}
> ```
> 
> It would be ideal if we could remove those duplicated `namespace` and `name` fields.

This is not possible a priori because one comes from the context and the other from the object. We could do it by not passing the object to the `debug`, `info`, `whatever` helper, but this is out of scope for this PR. There is another issue related to what to log: #189

File: internal/manager/logging/encoding.go [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	defaultDevEncodeLevel = zapcore.CapitalColorLevelEncoder
+	// the logger name is capitalized
+	defaultDevEncodeName = func(loggerName string, enc zapcore.PrimitiveArrayEncoder) {
+		enc.AppendString(strings.ToUpper(loggerName))
+	}
+)
+
+// setupLogEncoder sets some configurations proper to the development mode on the zap.Options.
+// In this way, the log structure is lighter and more human-friendly when the development mode
+// is enabled.
```
Comment (pmalek) [inline internal/manager/logging/encoding.go:32]:
```suggestion
// SetupLogEncoder sets additional logger configuration options when development mode is enabled.
// In this way, the log structure is lighter and more human-friendly when the development mode
// is enabled.
```

File: controllers/gateway_controller.go [hunk 3]
```diff
@@ -68,7 +69,14 @@ func (r *GatewayReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	log := log.FromContext(ctx).WithName("gateway")
+	var log logr.Logger
+	// if development mode is active, do not add the context to the log line, as we want
+	// to have a lighter logging structure
+	if r.DevelopmentMode {
+		log = ctrlruntimelog.Log.WithName("gateway")
+	} else {
+		log = ctrlruntimelog.FromContext(ctx).WithName("gateway")
+	}
```
Comment (pmalek) [inline controllers/gateway_controller.go:79]: This looks pretty much identical to what's in both `ControlPlaneReconciler` and `DataPlaneReconciler`. Don't you think that we could factor this out to a `func` and reuse it?

Comment (mlavacca) [reply controllers/gateway_controller.go:79]: sure

**Review by pmalek — APPROVED** [review]

**Pull request merged**